# CUDA Device Setup for Faster Training

This notebook demonstrates how to set up and use CUDA (GPU) for faster model training in PyTorch. Using GPU can speed up training by 10-100x compared to CPU, especially for deep learning models.

In [2]:
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np

# Check if CUDA is available
print("CUDA Available:", torch.cuda.is_available())
print("CUDA Version:", torch.version.cuda)
print("PyTorch Version:", torch.__version__)

if torch.cuda.is_available():
    print("GPU Device Count:", torch.cuda.device_count())
    print("Current GPU Device:", torch.cuda.current_device())
    print("GPU Device Name:", torch.cuda.get_device_name(0))
    print("GPU Memory Total:", torch.cuda.get_device_properties(0).total_memory / 1024**3, "GB")
else:
    print("CUDA is not available. Please check your GPU drivers and PyTorch installation.")

CUDA Available: True
CUDA Version: 11.8
PyTorch Version: 2.7.1+cu118
GPU Device Count: 1
Current GPU Device: 0
GPU Device Name: NVIDIA GeForce RTX 3080
GPU Memory Total: 11.99951171875 GB


In [7]:
# Set up device (CUDA if available, otherwise CPU)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Using device: {device}")

# If using GPU, print additional info
if device.type == 'cuda':
    print(f"GPU Memory Available: {torch.cuda.get_device_properties(0).total_memory / 1024**3:.2f} GB")
    print(f"GPU Memory Allocated: {torch.cuda.memory_allocated() / 1024**3:.2f} GB")
    print(f"GPU Memory Cached: {torch.cuda.memory_reserved() / 1024**3:.2f} GB")

Using device: cpu


# ✅ CUDA Hardware Available - Need PyTorch CUDA Installation

Your system has:
- ✅ NVIDIA GeForce RTX 3080 GPU
- ✅ NVIDIA drivers (version 580.88)
- ✅ CUDA runtime (version 13.0)

**Problem**: Your notebook is using PyTorch CPU version instead of CUDA version.

## Quick Fix - Install CUDA-enabled PyTorch

**Option 1 (Recommended): Use pip with CUDA wheels**
```powershell
# In PowerShell/Command Prompt:
pip uninstall torch torchvision torchaudio
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
```

**Option 2: Use conda (if you have Anaconda/Miniconda)**
```powershell
conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia
```

## After Installation
1. **Restart your VS Code completely** (close and reopen)
2. Make sure VS Code uses the correct Python interpreter:
   - Press `Ctrl+Shift+P`
   - Type "Python: Select Interpreter" 
   - Choose the interpreter where you installed PyTorch with CUDA
3. Re-run the first cell to verify CUDA is working

## Alternative: Use Windows Store Python (Already Has CUDA PyTorch)
Your system already has PyTorch with CUDA installed under:
`C:\Users\PC\AppData\Local\Microsoft\WindowsApps\python.exe`

You can switch to this interpreter:
- Press `Ctrl+Shift+P` → "Python: Select Interpreter"
- Choose the interpreter path that contains "WindowsApps"

In [6]:
# 🔍 Quick Verification - Check Python Interpreter and CUDA Status
import sys
import torch

print("=" * 60)
print("PYTHON INTERPRETER VERIFICATION")
print("=" * 60)
print(f"Python executable: {sys.executable}")
print(f"Python version: {sys.version.split()[0]}")

print("\n" + "=" * 60)
print("PYTORCH & CUDA STATUS")
print("=" * 60)
print(f"PyTorch version: {torch.__version__}")
print(f"CUDA compiled version: {torch.version.cuda}")
print(f"CUDA available: {torch.cuda.is_available()}")

if torch.cuda.is_available():
    print(f"CUDA device count: {torch.cuda.device_count()}")
    print(f"Current CUDA device: {torch.cuda.current_device()}")
    print(f"CUDA device name: {torch.cuda.get_device_name(0)}")
    props = torch.cuda.get_device_properties(0)
    print(f"CUDA device memory: {props.total_memory / 1024**3:.2f} GB")
    
    # Quick GPU test
    print("\n" + "=" * 60)
    print("CUDA FUNCTIONALITY TEST")
    print("=" * 60)
    try:
        x = torch.randn(1000, 1000, device='cuda')
        y = torch.randn(1000, 1000, device='cuda') 
        z = torch.matmul(x, y)
        print("✅ CUDA matrix multiplication successful!")
        print(f"✅ Result tensor device: {z.device}")
        print(f"✅ GPU memory allocated: {torch.cuda.memory_allocated() / 1024**2:.1f} MB")
    except Exception as e:
        print(f"❌ CUDA test failed: {e}")
else:
    print("\n❌ CUDA is not available")
    print("👉 Follow the installation steps above")
    
print("\n" + "=" * 60)

PYTHON INTERPRETER VERIFICATION
Python executable: c:\Users\PC\AppData\Local\Programs\Python\Python313\python.exe
Python version: 3.13.5

PYTORCH & CUDA STATUS
PyTorch version: 2.7.1+cpu
CUDA compiled version: None
CUDA available: False

❌ CUDA is not available
👉 Follow the installation steps above



## 🔧 **SOLUTION: Kernel vs Terminal Python Mismatch**

**Problem Identified**: 
- ✅ You successfully installed PyTorch+CUDA in terminal
- ❌ VS Code notebook is using a different Python interpreter (Python 3.13)
- ❌ The notebook kernel needs to be updated

## 🚀 **Quick Fix Steps:**

### **Step 1: Change Python Interpreter**
1. Press `Ctrl+Shift+P`
2. Type "Python: Select Interpreter"
3. Choose the Python that matches your terminal (look for Python 3.11 or the path where you installed PyTorch)

### **Step 2: Restart Kernel**
1. Press `Ctrl+Shift+P`
2. Type "Python: Restart Kernel"
3. Wait for kernel to restart

### **Step 3: Test Again**
- Re-run the verification cell above
- You should see CUDA Available: True

## 🔍 **Alternative: Install in Current Environment**
If you prefer to keep using Python 3.13, run the installation cell below:

## 🔄 **RESTART KERNEL NOW**

**Installation completed successfully!** 

To activate CUDA, you MUST restart the kernel:

### **RESTART STEPS:**
1. Press `Ctrl+Shift+P`
2. Type "**Python: Restart Kernel**"
3. Click on it
4. Wait for "Kernel Restarted" message
5. Re-run the verification cell above

**Or use the VS Code menu:**
- Go to "View" → "Command Palette" 
- Search "Python: Restart Kernel"

In [5]:
# 🚀 QUICK FIX: Install CUDA PyTorch in Current Environment
# Run this cell to install PyTorch with CUDA support

import subprocess
import sys

print("Installing PyTorch with CUDA support...")
print("This may take a few minutes...")

try:
    # Uninstall CPU version
    subprocess.check_call([
        sys.executable, "-m", "pip", "uninstall", 
        "torch", "torchvision", "torchaudio", "-y"
    ])
    
    # Install CUDA version
    subprocess.check_call([
        sys.executable, "-m", "pip", "install", 
        "torch", "torchvision", "torchaudio", 
        "--index-url", "https://download.pytorch.org/whl/cu118"
    ])
    
    print("✅ Installation completed!")
    print("🔄 Please restart the kernel: Ctrl+Shift+P -> 'Python: Restart Kernel'")
    print("📝 Then re-run the verification cell above")
    
except subprocess.CalledProcessError as e:
    print(f"❌ Installation failed: {e}")
    print("💡 Try running these commands in PowerShell:")
    print("   pip uninstall torch torchvision torchaudio -y")
    print("   pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118")

Installing PyTorch with CUDA support...
This may take a few minutes...
✅ Installation completed!
🔄 Please restart the kernel: Ctrl+Shift+P -> 'Python: Restart Kernel'
📝 Then re-run the verification cell above
✅ Installation completed!
🔄 Please restart the kernel: Ctrl+Shift+P -> 'Python: Restart Kernel'
📝 Then re-run the verification cell above


In [1]:
# After installing CUDA-enabled PyTorch, restart the kernel and run this cell
# You can restart kernel with: Ctrl+Shift+P -> "Python: Restart Kernel"

import torch
print("=== CUDA Status After Installation ===")
print(f"CUDA Available: {torch.cuda.is_available()}")
print(f"PyTorch Version: {torch.__version__}")
print(f"CUDA Version: {torch.version.cuda}")

if torch.cuda.is_available():
    print(f"GPU Device: {torch.cuda.get_device_name(0)}")
    print(f"GPU Memory: {torch.cuda.get_device_properties(0).total_memory / 1024**3:.2f} GB")
    print("✅ CUDA is ready for use!")
else:
    print("❌ CUDA is still not available")
    print("Try restarting VS Code completely")

=== CUDA Status After Installation ===
CUDA Available: True
PyTorch Version: 2.7.1+cu118
CUDA Version: 11.8
GPU Device: NVIDIA GeForce RTX 3080
GPU Memory: 12.00 GB
✅ CUDA is ready for use!


In [4]:
# Example: How to move your model and data to GPU
# This is how you should modify your existing training code

# 1. Move your model to GPU
# model = model.to(device)  # Move model to GPU

# 2. Move your data to GPU (in your training loop)
# inputs, labels = inputs.to(device), labels.to(device)

# Example with a simple CNN model
class SimpleCNN(nn.Module):
    def __init__(self, num_classes=10):
        super(SimpleCNN, self).__init__()
        self.conv1 = nn.Conv2d(3, 32, 3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, 3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.fc1 = nn.Linear(64 * 8 * 8, 128)
        self.fc2 = nn.Linear(128, num_classes)
        self.dropout = nn.Dropout(0.5)
        
    def forward(self, x):
        x = self.pool(torch.relu(self.conv1(x)))
        x = self.pool(torch.relu(self.conv2(x)))
        x = x.view(-1, 64 * 8 * 8)
        x = torch.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

# Create model and move to GPU
model = SimpleCNN(num_classes=10)
model = model.to(device)  # This moves the model to GPU
print(f"Model is on device: {next(model.parameters()).device}")

# Example data (batch_size=32, channels=3, height=32, width=32)
sample_input = torch.randn(32, 3, 32, 32).to(device)  # Move data to GPU
sample_labels = torch.randint(0, 10, (32,)).to(device)  # Move labels to GPU

print(f"Sample input is on device: {sample_input.device}")
print(f"Sample labels are on device: {sample_labels.device}")

Model is on device: cuda:0
Sample input is on device: cuda:0
Sample labels are on device: cuda:0


In [5]:
# Example training loop with GPU
def train_step_example():
    # Set model to training mode
    model.train()
    
    # Define loss function and optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)
    
    # Example training step
    optimizer.zero_grad()
    
    # Forward pass (data is already on GPU)
    outputs = model(sample_input)
    loss = criterion(outputs, sample_labels)
    
    # Backward pass
    loss.backward()
    optimizer.step()
    
    print(f"Training loss: {loss.item():.4f}")
    print(f"Loss tensor is on device: {loss.device}")
    
    return loss.item()

# Run example training step
loss_value = train_step_example()

Training loss: 2.3383
Loss tensor is on device: cuda:0
